File: src/main/java/org/cloudfoundry/credhub/repository/PermissionRepository.java
package org.cloudfoundry.credhub.repository;


import org.cloudfoundry.credhub.entity.PermissionData;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

public interface PermissionRepository extends JpaRepository<PermissionData, UUID> {
  List<PermissionData> findAllByPath(String path);

  List<PermissionData> findAllByActor(String actor);

  PermissionData findByPathAndActor(String path, String actor);

  PermissionData findByUuid(UUID uuid);

  @Query(value = "select * from permission where path IN ?1 AND actor=?2",  nativeQuery = true)
  List<PermissionData> findByPathsAndActor(List<String> paths, String actor);

  @Query(value = "select path from permission where read_permission = TRUE and actor = ?1",  nativeQuery = true)
  List<String> findAllPathsForActorWithReadPermission(String actor);

  @Transactional
  long deleteByPathAndActor(String path, String actor);
}

File: components/permissions/src/main/kotlin/org/cloudfoundry/credhub/repositories/PermissionRepository.kt
package org.cloudfoundry.credhub.repositories

import org.cloudfoundry.credhub.data.PermissionData
import org.springframework.context.annotation.Profile
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@Profile("!stub-repositories")
interface PermissionRepository : JpaRepository<PermissionData, UUID> {
    fun findAllByPath(path: String?): List<PermissionData>

    fun findByPathAndActor(
        path: String,
        actor: String,
    ): PermissionData?

    fun findByUuid(uuid: UUID): PermissionData?

    @Query(value = "select * from permission where path IN ?1 AND actor=?2", nativeQuery = true)
    fun findByPathsAndActor(
        paths: List<String>,
        actor: String,
    ): List<PermissionData>

    @Query(value = "select path from permission where read_permission = TRUE and actor = ?1", nativeQuery = true)
    fun findAllPathsForActorWithReadPermission(actor: String): List<String>

    @Transactional
    fun deleteByPathAndActor(
        path: String,
        actor: String,
    ): Long
}
